Fix SFTPSensor.newer_than not working with jinja logical ds/ts expression

grrolland opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

closes: #36629 

Add str to time transformation in case the newer_than parameter is templated.


<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
Fixes https://github.com/apache/airflow/issues/36629

Commit:
Fixes https://github.com/apache/airflow/issues/36629

hussein-awala committed:
update simple-salesforce type hints to support 1.12.6 (#39047)

Commit:
Fixes PR failed test

Commit:
Fixes PR failed test

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating

ephraimbuddy committed:
Add changelog for airflow python client 2.9.0 (#39060)

potiuk committed:
Upgrade to latest hatchling as build dependency (#39044)

eladkal committed:
Prepare docs 1st wave (RC3) + ad hoc April 2024 (#38995) (#39054)

* Prepare docs 1st wave (RC3) + ad hoc April 2024 (#38995)

* update databricks

rawwar committed:
[docs] update `DagBag` class docstring to include all params (#38814)

* update docstring for DagBag class

* break long line

* fix space

Signed-off-by: kalyanr <kalyan.ben10@live.com>

---------

Signed-off-by: kalyanr <kalyan.ben10@live.com>

lzdanski committed:
Data aware scheduling docs edits (#38687)

potiuk committed:
Moves airflow import in deprecated pod_generator to local (#39062)

The import might be invoked when K8S executor starts with sentry on
and it might lead to circular imports

Related: #31442

dirrao commented on tests/providers/sftp/sensors/test_sftp.py:
```diff
@@ -108,6 +108,17 @@ def test_naive_datetime(self, sftp_hook_mock):
         sftp_hook_mock.return_value.get_mod_time.assert_called_once_with("/path/to/file/1970-01-01.txt")
         assert not output
 
+    @patch("airflow.providers.sftp.sensors.sftp.SFTPHook")
```
I would suggest to parameterize the newer_than rather than reapting the same code.

grrolland replied:
Your are right, my bad. It's done now.

jedcunningham committed:
KPO xcom sidecar PodDefault usage (#38951)

We should use the same, non deprecated, version of PodDefaults for the
xcom sidecar when creating and reading xcom.

Commit:
Remove an parametrize duplicate tests

Commit:
Remove an parametrize duplicate tests

Commit:
Merge remote-tracking branch 'origin/fix-36629-sftp-sensor-newer-than-templating' into fix-36629-sftp-sensor-newer-than-templating

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating

vincbeck committed:
Add examples in AWS auth manager documentation (#39040)

humit0 committed:
update document (#39068)

ephraimbuddy committed:
Update hatchling to version 1.24.0 (#39072)

bbovenzi committed:
Check that the dataset<>task exists before trying to render graph (#39069)

Commit:
Fix formatting

Commit:
Fix formatting

Commit:
Merge remote-tracking branch 'origin/fix-36629-sftp-sensor-newer-than-templating' into fix-36629-sftp-sensor-newer-than-templating

Commit:
Fix formatting

Commit:
Fix formatting

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating

Taragolis commented on tests/providers/sftp/sensors/test_sftp.py:
```diff
@@ -97,11 +97,12 @@ def test_file_not_new_enough(self, sftp_hook_mock):
         sftp_hook_mock.return_value.get_mod_time.assert_called_once_with("/path/to/file/1970-01-01.txt")
         assert not output
 
+    @pytest.mark.parametrize("newer_than", (datetime(2020, 1, 2), "2020-01-02 00:00:00+00:00"))
```
Small nit, "2020-01-02 00:00:00+00:00" - doesn't look as naive datetime

grrolland replied:
Doesn't look but ... It's actually the str representation of "datetime(2020, 1, 2)" like it could be passed as a jina template in newer_than.

Taragolis replied:
- datetime(2020, 1, 2) parsed as a tz-naive datetime, and required to convert it to the aware again, which is not as trivial as it might seem
- `"2020-01-02 00:00:00+00:00"` as tz-aware datetime and already has timezone information

`datetime(2020, 1, 2) != datetime(2020, 1, 2, tzinfo=timezone.utc)`

grrolland replied:
Rename and add test case parameters.

Taragolis commented on airflow/providers/sftp/sensors/sftp.py:
```diff
@@ -21,6 +21,7 @@
 
 import os
 from datetime import datetime, timedelta
+from dateutil.parser import parse as parse_date
```
Is it actually required to parse ds/ts which are ISO format representation? Because if a target to parse ISO format (other formats might be not parsed correctly) you could just use a `datetime.datetime.fromisoformat(...)`

grrolland replied:
newer_than accept originaly a datetime type. The meaning of the PR is to accept the same type in a jinja template. 
ds/ts are, you are right in ISO format. 
But if you try this : 
```
from datetime import datetime
from dateutil.parser import parse as parse_date


print(datetime(2020, 1, 2))
print(parse_date("2020-01-02 00:00:00"))
print(parse_date("2018-01-01T00:00:00+00:00"))
print(parse_date("2020-01-02T00:00:00Z"))
```
It will render : 
```
2020-01-02 00:00:00
2020-01-02 00:00:00
2018-01-01 00:00:00+00:00
2020-01-02 00:00:00+00:00

```

grrolland replied:
Note that the same methode is used in the SFTPHook in the deferred mode. I use this method to be coherent with this.

grrolland replied:
Sorry, not the SFTPHook, the SFTPTrigger.

Taragolis replied:
> newer_than accept originaly a datetime type. The meaning of the PR is to accept the same type in a jinja template.
ds/ts are, you are right in ISO format. But if you try this :

Don't get this example, could you elaborate a bit more?

> Note that the same methode is used in the SFTPHook in the deferred mode. I use this method to be coherent with this.

This not a point if some part already use it, it might be add years ago, or by a mistake.

Taragolis replied:
My point here why we need yet another way to parse ISO format, or parse datetime/dates if we already have [`airflow.utils.timezone.parse`](https://github.com/apache/airflow/blob/0a74928894fb57b0160208262ccacad12da23fc7/airflow/utils/timezone.py#L194-L204)

grrolland replied:
Use [airflow.utils.timezone.parse](https://github.com/apache/airflow/blob/0a74928894fb57b0160208262ccacad12da23fc7/airflow/utils/timezone.py#L194-L204) now.

Commit:
Change date/time parsing method for newer_than parameter un SFTPSensor

Commit:
Merge remote-tracking branch 'origin/fix-36629-sftp-sensor-newer-than-templating' into fix-36629-sftp-sensor-newer-than-templating

# Conflicts:
#	airflow/providers/sftp/sensors/sftp.py
#	tests/providers/sftp/sensors/test_sftp.py

Commit:
Change date/time parsing method for newer_than parameter un SFTPSensor

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating

Taragolis commented on tests/providers/sftp/sensors/test_sftp.py:
```diff
@@ -97,11 +97,24 @@ def test_file_not_new_enough(self, sftp_hook_mock):
+    @pytest.mark.parametrize(
+        "newer_than", (
+            datetime(2020, 1, 2),
+            datetime(2020, 1, 2, tzinfo=timezone.utc),
```
Test failed due to `pendulum.timezone` is a function, I guess you want to use here `datetime.timezone.utc` here

So just import it and use here

```python
...
from datetime import timezone as stdlib_timezone
...
            datetime(2020, 1, 2, tzinfo=stdlib_timezone.utc),

```

Commit:
Fix utc timezone in unit tests

Commit:
Merge remote-tracking branch 'origin/fix-36629-sftp-sensor-newer-than-templating' into fix-36629-sftp-sensor-newer-than-templating

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating

Commit:
Fix utc timezone in unit tests

Commit:
Merge remote-tracking branch 'origin/fix-36629-sftp-sensor-newer-than-templating' into fix-36629-sftp-sensor-newer-than-templating

grrolland committed:
Merge branch 'main' into fix-36629-sftp-sensor-newer-than-templating
